Mb db connection uri parsing

dpsutton opened the pull request:
fixes https://github.com/metabase/metabase/issues/14678

Using the raw jdbc string when its of the form
`username:password@host:port` causes problems if passed directly to
jdbc. Clojure.java.jdbc has some special handling for username and
passwords

```clojure
(if-let [user-info (.getUserInfo uri)]
  {:user (first (str/split user-info #":"))
   :password (second (str/split user-info #":"))})
```

Our logic here is:
1. if mb-db-connection-uri is set and not inline credentials,
return that, possibly prepending jdbc: to it
2. if mb-db-connection-uri is set and is using inline credentials, then
we use the old parsing code which doesn't support all of the crazy
stuff individual connection strings may offer and return a db-spec
3. if all the individual parts are supplied (mb-db-host, mb-db-port,
etc) are used to construct a db-spec

dpsutton committed:
Parse mb-db-connection-uri when it contains a password

Using the raw jdbc string when its of the form
`username:password@host:port` causes problems if passed directly to
jdbc. Clojure.java.jdbc has some special handling for username and
passwords

```clojure
(if-let [user-info (.getUserInfo uri)]
  {:user (first (str/split user-info #":"))
   :password (second (str/split user-info #":"))})
```

Our logic here is:
1. if mb-db-connection-uri is set and not an older style password,
return that, possibly prepending jdbc: to it
2. if mb-db-connection-uri is set and is an older style passowrd, then
we use the old parsing code which doesn't support all of the crazy
stuff individual connection strings may offer and return a db-spec
3. if all the individual parts are supplied (mb-db-host, mb-db-port,
etc) are used to construct a db-spec

dpsutton committed:
Correct what old style means for conn-uri

accidentally thought user@host/metabase?password=password was ok. Any
credentially in the host spot is not supported

dpsutton committed:
Cleanup documentation

don't refer to old password style but call it inline credentials

kidd reviewed (approved):
lgtm.
kicked the tests again

flamber reviewed (commented):
Works with both `user:pass@host` and `user@host?password=pass` :+1: 

But if I use `postgres://` instead of `postgresql://`, then it stalls - example:
`postgres://172.17.0.3:5432/mb?user=postgres&password=password`
Errors with:
`Caused by: java.sql.SQLException: No suitable driver found for jdbc:postgres://172.17.0.3:5432/mb?user=postgres&password=password`
It has been used in [docs](https://github.com/metabase/metabase/blob/release-x.38.x/docs/operations-guide/configuring-application-database.md#postgres) and forum for many years, so don't think we can get rid of `postgres` as synonym of `postgresql`

Micro minor: Can the deprecation warning get a red color?

dpsutton committed:
Fixups for connection parsing

- ensure we turn postgres: to postgresql: in connection string
- ensure we warn about postgres ssl edge case when using raw
  connection string
- make warnings testable

dpsutton committed:
Space in inline credentials warning

dpsutton commented:
@flamber they are nice and red.

Also, I made sure we are logging about the postgres ssl issue #8908 when we use the connection string raw. Added some tests relating to the warnings we emit as well.

dpsutton committed:
Strip jdbc from connection when checking if inline credentials

flamber commented:
@dpsutton Do you need me to do another round of testing or are you all good?

dpsutton committed:
Fix for old credential style

dpsutton commented:
@flamber probably couldn't hurt to pass again. if this is incorrect it could completely stop the application.

flamber reviewed (approved):
Tested `ca4015a` and cannot stall it in any way on my local setup :+1:

camsaul reviewed (approved):
LGTM. Try merging in latest `release-x.38.x`, I tweaked max heap size on Presto which should fix the random test failures there
